File: mobile/lib/services/auth_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../core/constants/api_constants.dart';
import '../models/user_model.dart';
import 'api_service.dart';

class AuthService {
  final ApiService _api;

  AuthService(this._api);

  /// Login → simpan token ke SharedPreferences → return UserModel.
  /// Lempar String pesan error jika gagal.
  Future<UserModel> login(String email, String password) async {
    try {
      final response = await _api.post(ApiConstants.login, {
        'email': email,
        'password': password,
      });
      final body = response.data as Map<String, dynamic>;
      final user = UserModel.fromJson(body['data'] as Map<String, dynamic>);
      final token = body['token'] as String;

      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(ApiConstants.tokenKey, token);

      return user;
    } on DioException catch (e) {
      throw _pesanError(e, 'Login gagal. Periksa email dan password Anda.');
    }
  }

  /// Register → lempar String pesan error jika gagal.
  Future<void> register(
    String username,
    String email,
    String password,
  ) async {
    try {
      await _api.post(ApiConstants.register, {
        'username': username,
        'email': email,
        'password': password,
      });
    } on DioException catch (e) {
      throw _pesanError(e, 'Registrasi gagal. Coba lagi nanti.');
    }
  }

  /// Hapus token dari SharedPreferences.
  Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(ApiConstants.tokenKey);
  }

  /// Kembalikan true jika token ada dan tidak kosong.
  Future<bool> isLoggedIn() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString(ApiConstants.tokenKey);
    return token != null && token.isNotEmpty;
  }

  /// Kembalikan token tersimpan (null jika belum login).
  Future<String?> getSavedToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(ApiConstants.tokenKey);
  }

  String _pesanError(DioException e, String fallback) {
    final data = e.response?.data;
    if (data is Map<String, dynamic>) {
      return data['pesan'] as String? ?? fallback;
    }
    if (e.type == DioExceptionType.connectionTimeout ||
        e.type == DioExceptionType.receiveTimeout) {
      return 'Koneksi ke server timeout. Pastikan server berjalan.';
    }
    if (e.type == DioExceptionType.connectionError) {
      return 'Tidak dapat terhubung ke server. Periksa koneksi jaringan.';
    }
    return fallback;
  }
}

final authServiceProvider = Provider<AuthService>((ref) {
  return AuthService(ref.read(apiServiceProvider));
});

File: mobile-rn/src/services/authService.js
import * as SecureStore from 'expo-secure-store';
import api from './api';
import { API } from '../constants/api';

function pesanError(error, fallback) {
  const data = error.response?.data;
  if (data?.pesan) return data.pesan;
  if (error.code === 'ECONNABORTED') return 'Koneksi timeout. Pastikan server berjalan.';
  if (error.message === 'Network Error') return 'Tidak dapat terhubung ke server.';
  return fallback;
}

export async function login(email, password) {
  try {
    const res = await api.post('/api/auth/login', { email, password });
    const { token, data: user } = res.data;
    await SecureStore.setItemAsync(API.TOKEN_KEY, token);
    await SecureStore.setItemAsync('user_data', JSON.stringify(user));
    return { user, token };
  } catch (error) {
    throw pesanError(error, 'Login gagal. Periksa email dan password.');
  }
}

export async function getSavedUser() {
  const raw = await SecureStore.getItemAsync('user_data');
  return raw ? JSON.parse(raw) : null;
}

export async function register(username, email, password) {
  try {
    await api.post('/api/auth/register', { username, email, password });
  } catch (error) {
    throw pesanError(error, 'Registrasi gagal. Coba lagi nanti.');
  }
}

export async function logout() {
  await SecureStore.deleteItemAsync(API.TOKEN_KEY);
  await SecureStore.deleteItemAsync('user_data');
}

export async function getSavedToken() {
  return SecureStore.getItemAsync(API.TOKEN_KEY);
}
